File: app/core/bb84.js
// BB84 quantum key distribution for the Tract web client — the same ceremony as
// the Go reference (internal/quantum/bb84.go) and the iOS port (QuantumBB84.swift),
// wire-compatible so a web↔native call agrees bit-for-bit. BB84 lets the two call
// participants agree on a shared key AND detect tampering: measuring an unknown
// qubit in the wrong basis disturbs it (no-cloning), surfacing as an elevated error
// rate (QBER) on a publicly compared sample. If QBER crosses the abort threshold we
// assume interception and THROW THE KEY AWAY — the call collapses rather than
// continuing compromised.
//
// Honest scope: software can't put a real photon on the wire, so this isn't the
// information-theoretic secrecy of physical QKD. It's a faithful BB84 *ceremony*
// over the already-authenticated/E2E signaling channel — an active man-in-the-
// middle who perturbs the exchange (intercept-resend) is caught by the QBER check.

// Sampled error rate above which we assume compromise. Honest channels sit near 0;
// intercept-resend drives QBER toward 25%, so 15% cleanly separates the two. Must
// match QBERAbortThreshold in bb84.go / kQBERAbortThreshold in QuantumBB84.swift.
export const QBER_ABORT_THRESHOLD = 0.15;

const RECTILINEAR = 0; // Z basis
const DIAGONAL = 1;    // X basis

// Classical decisions (which bit, which basis) come from the CSPRNG, exactly as a
// real QKD device chooses them locally; only the qubit carries quantum behaviour.
function randFloat() {
  const buf = new Uint32Array(2);
  crypto.getRandomValues(buf);
  // 53-bit uniform in [0,1).
  return (buf[0] * 0x100000000 + buf[1]) / 0x20000000000000;
}
function randBit() { return randFloat() < 0.5 ? 0 : 1; }
function randBasis() { return randBit() === 1 ? DIAGONAL : RECTILINEAR; }

// --- Single-qubit state-vector emulator (all BB84 needs) ----------------------
// State a|0> + b|1> as complex amplitudes [a0r,a0i,a1r,a1i].
function zero() { return [1, 0, 0, 0]; }
function applyX(s) { return [s[2], s[3], s[0], s[1]]; }
function applyH(s) {
  const r = Math.SQRT1_2;
  return [(s[0] + s[2]) * r, (s[1] + s[3]) * r, (s[0] - s[2]) * r, (s[1] - s[3]) * r];
}
// Measure in computational (Z) basis, collapsing. Entropy from the CSPRNG — an
// honest emulation, like the Go qrng.
function measureZ(s) {
  const p1 = s[2] * s[2] + s[3] * s[3];
  return randFloat() < p1 ? 1 : 0;
}

function prepareQubit(bit, basis) {
  let s = zero();
  if (bit === 1) s = applyX(s);
  if (basis === DIAGONAL) s = applyH(s);
  return s;
}
function measureQubit(s, basis) {
  if (basis === DIAGONAL) s = applyH(s);
  return measureZ(s);
}

// Derive a 32-byte key from sifted bits (SHA-256 over the packed bits). Bit-packing
// matches DeriveKey in Go and deriveKey in Swift.
async function deriveKey(bits) {
  const packed = new Uint8Array(Math.ceil(bits.length / 8));
  bits.forEach((b, i) => { if (b === 1) packed[i >> 3] |= 1 << (i & 7); });
  const digest = await crypto.subtle.digest('SHA-256', packed);
  return new Uint8Array(digest);
}

/**
 * Drives a BB84 ceremony between two peers over the existing signaling channel.
 * The connection initiator plays Alice (prepares qubits); the other peer plays Bob
 * (measures them). Messages travel as signaling type "qkd" with a JSON payload
 * keyed by `p` (phase):
 *   p="q"  Alice→Bob : prepared qubits           {n, q:[[a0r,a0i,a1r,a1i],…]}
 *   p="b"  Bob→Alice : Bob's measurement bases    {bb:[0/1,…]}
 *   p="s"  Alice→Bob : Alice's bases + QBER sample {ab:[0/1,…], sb:[bit,…]}
 *   p="ok" Bob→Alice : QBER acceptable, key agreed {}
 *   p="x"  Bob→Alice : QBER over threshold         {}
 * Mirrors reconcile() in bb84.go and BB84Session in the iOS app.
 */
export class BB84Session {
  static QUBIT_COUNT = 96;

  /**
   * @param {'alice'|'bob'} role
   * @param {(msg: object) => void} send  routes one ceremony message as type "qkd"
   * @param {(key: Uint8Array) => void} onSuccess
   * @param {(reason: string) => void} onAbort
   */
  constructor(role, send, onSuccess, onAbort) {
    this.role = role;
    this.send = send;
    this.onSuccess = onSuccess;
    this.onAbort = onAbort;
    this.finished = false;
    this.aBits = []; this.aBases = [];
    this.bBits = []; this.bBases = [];
  }

  start() {
    if (this.role !== 'alice') return;
    const n = BB84Session.QUBIT_COUNT;
    this.aBits = Array.from({ length: n }, () => randBit());
    this.aBases = Array.from({ length: n }, () => randBasis());
    const q = this.aBits.map((bit, i) => prepareQubit(bit, this.aBases[i]));
    this.send({ p: 'q', n, q });
  }

  async handle(payload) {
    if (this.finished || !payload || typeof payload.p !== 'string') return;
    const p = payload.p;
    if (this.role === 'bob' && p === 'q') return this._bobReceiveQubits(payload);
    if (this.role === 'alice' && p === 'b') return this._aliceReceiveBases(payload);
    if (this.role === 'bob' && p === 's') return this._bobReconcile(payload);
    if (this.role === 'alice' && p === 'ok') return this._aliceFinish(true);
    if (this.role === 'alice' && p === 'x') return this._aliceFinish(false);
  }

  // --- Bob ---
  _bobReceiveQubits(payload) {
    const raw = Array.isArray(payload.q) ? payload.q : [];
    this.bBits = []; this.bBases = [];
    for (const amp of raw) {
      if (!Array.isArray(amp) || amp.length !== 4) continue;
      const basis = randBasis();
      this.bBases.push(basis);
      this.bBits.push(measureQubit(amp.slice(), basis));
    }
    this.send({ p: 'b', bb: this.bBases });
  }

  async _bobReconcile(payload) {
    const ab = payload.ab, sampleA = payload.sb;
    if (!Array.isArray(ab) || !Array.isArray(sampleA) || ab.length !== this.bBases.length) {
      this.finished = true; this.onAbort('BB84 protocol error'); return;
    }
    const sifted = [];
    for (let i = 0; i < ab.length; i++) if (ab[i] === this.bBases[i]) sifted.push(this.bBits[i]);
    if (sifted.length === 0) return this._abortBob('BB84: no sifted bits');

    const sampleB = [], keyBits = [];
    sifted.forEach((bit, i) => (i % 2 === 0 ? sampleB : keyBits).push(bit));
    const m = Math.min(sampleA.length, sampleB.length);
    let mism = 0;
    for (let i = 0; i < m; i++) if (sampleA[i] !== sampleB[i]) mism++;
    const qber = m > 0 ? mism / m : 0;

    if (qber > QBER_ABORT_THRESHOLD || keyBits.length === 0) {
      return this._abortBob(`Канал скомпрометирован (QBER ${Math.round(qber * 100)}%)`);
    }
    this.finished = true;
    this.send({ p: 'ok' });
    this.onSuccess(await deriveKey(keyBits));
  }

  _abortBob(reason) {
    this.finished = true;
    this.send({ p: 'x' });
    this.onAbort(reason);
  }

  // --- Alice ---
  _aliceReceiveBases(payload) {
    const bb = payload.bb;
    if (!Array.isArray(bb) || bb.length !== this.aBases.length) return;
    this.bBases = bb;
    const sifted = [];
    for (let i = 0; i < this.aBases.length; i++) if (this.aBases[i] === bb[i]) sifted.push(this.aBits[i]);
    const sample = sifted.filter((_, i) => i % 2 === 0);
    this.send({ p: 's', ab: this.aBases, sb: sample });
  }

  async _aliceFinish(success) {
    this.finished = true;
    if (!success) { this.onAbort('Канал скомпрометирован (BB84)'); return; }
    const sifted = [];
    for (let i = 0; i < this.aBases.length; i++) if (this.aBases[i] === this.bBases[i]) sifted.push(this.aBits[i]);
    const keyBits = sifted.filter((_, i) => i % 2 === 1);
    this.onSuccess(await deriveKey(keyBits));
  }
}

File: internal/quantum/bb84.go
package quantum

// BB84 quantum key distribution, run on the state-vector emulator. BB84 lets two
// parties agree on a shared secret key AND detect eavesdropping: any attempt to
// learn the qubits in transit disturbs them (you cannot measure an unknown qubit
// without, in general, perturbing it — the no-cloning theorem), which shows up as
// an elevated error rate (QBER) on a publicly compared sample. If the QBER is over
// the abort threshold, the parties assume interception and THROW THE KEY AWAY — in
// Tract that means the call/chat collapses rather than continuing compromised.
//
// What this code is, honestly:
//   - A faithful BB84 protocol over a real (emulated) quantum channel, including
//     the intercept-resend attack and its ~25% QBER signature. The collapse on
//     measurement is genuine emulator behaviour, not a mock.
//   - Software cannot transmit a physical qubit over a classical network, so the
//     information-theoretic secrecy of real QKD requires real photons. Here BB84
//     serves as a tamper-evident key-agreement ceremony layered on the existing
//     authenticated/E2E channel: an active attacker who disturbs the exchange is
//     detected by the QBER check and the session is torn down.
//
// Bases: Rectilinear (Z, {|0>,|1>}) and Diagonal (X, {|+>,|->}).

import (
	"crypto/sha256"
)

// Basis is the measurement/preparation basis of a qubit.
type Basis int

const (
	Rectilinear Basis = 0 // Z basis: |0>, |1>
	Diagonal    Basis = 1 // X basis: |+>, |->
)

// QBERAbortThreshold is the sampled error rate above which we assume the channel
// is compromised and abort. Honest channels sit near 0; an intercept-resend
// eavesdropper drives QBER toward 25%, so 15% cleanly separates the two.
const QBERAbortThreshold = 0.15

// prepare returns a 1-qubit register encoding bit in the given basis.
//
//	Rectilinear: bit 0 -> |0>, bit 1 -> |1>
//	Diagonal:    bit 0 -> |+>, bit 1 -> |->
func prepare(bit int, basis Basis) *Register {
	r := New(1)
	if bit == 1 {
		r.X(0) // |0> -> |1>
	}
	if basis == Diagonal {
		r.H(0) // |0>->|+>, |1>->|->
	}
	return r
}

// measure measures the register in the given basis and returns the bit. Measuring
// in the diagonal basis is a Hadamard followed by a computational measurement.
func measure(r *Register, basis Basis) int {
	if basis == Diagonal {
		r.H(0)
	}
	return r.Measure(0)
}

// randBit returns a random classical bit. Alice's data bits and both parties'
// basis *choices* are classical decisions a real QKD device makes with a local
// RNG — they are never themselves transmitted as qubits — so we draw them from
// crypto/rand. The quantum emulator is reserved for the qubits that actually
// travel the channel (prepare/measure), where collapse-on-measurement is what
// makes eavesdropping detectable. (Sampling a bit via the emulator would mean
// allocating a fresh register per coin flip — correct but needlessly costly.)
func randBit() int {
	if randFloat() < 0.5 {
		return 0
	}
	return 1
}

func randBasis() Basis {
	if randBit() == 1 {
		return Diagonal
	}
	return Rectilinear
}

// Result is the outcome of a BB84 run.
type Result struct {
	Key       []byte  // derived shared key (empty if aborted)
	KeyBits   []int   // sifted+checked key bits both sides hold (empty if aborted)
	QBER      float64 // measured error rate on the sampled bits
	Sifted    int     // number of bits surviving basis sifting
	Sampled   int     // number of sifted bits sacrificed to estimate QBER
	Aborted   bool    // true if QBER exceeded the threshold (eavesdropper assumed)
	Agreement bool    // true if both parties' surviving key bits matched
}

// Eavesdropper models an intercept-resend attacker on the quantum channel: it
// measures each qubit in a basis it guesses, then resends a fresh qubit prepared
// from what it saw. When its basis guess is wrong it both learns nothing useful
// AND disturbs the state, producing detectable errors.
type Eavesdropper struct{ active bool }

// Eve returns an active intercept-resend eavesdropper.
func Eve() *Eavesdropper { return &Eavesdropper{active: true} }

func (e *Eavesdropper) tap(r *Register) *Register {
	if e == nil || !e.active {
		return r
	}
	g := randBasis()
	bit := measure(r, g) // collapses r
	return prepare(bit, g)
}

// Simulate runs a full BB84 exchange over n qubits between two honest parties,
// optionally with an eavesdropper on the line. It returns the agreed key and the
// QBER-based verdict. This is the in-process reference used by tests and demos;
// the live protocol (see the Sender/Receiver steps) exchanges the same data over
// the network.
func Simulate(n int, eve *Eavesdropper) Result {
	// Alice's random bits and bases.
	aBits := make([]int, n)
	aBases := make([]Basis, n)
	bBases := make([]Basis, n)
	bBits := make([]int, n)

	for i := 0; i < n; i++ {
		aBits[i] = randBit()
		aBases[i] = randBasis()
		bBases[i] = randBasis()

		q := prepare(aBits[i], aBases[i])
		q = eve.tap(q) // possible interception (collapses + resends)
		bBits[i] = measure(q, bBases[i])
	}

	return reconcile(aBits, aBases, bBits, bBases)
}

// reconcile performs sifting, QBER sampling and key derivation given both sides'
// bits and bases. Splitting it out keeps the network protocol and the in-process
// simulation sharing one implementation.
func reconcile(aBits []int, aBases []Basis, bBits []int, bBases []Basis) Result {
	// Sift: keep positions where the bases matched.
	var aSift, bSift []int
	for i := range aBits {
		if aBases[i] == bBases[i] {
			aSift = append(aSift, aBits[i])
			bSift = append(bSift, bBits[i])
		}
	}
	res := Result{Sifted: len(aSift)}
	if len(aSift) == 0 {
		res.Aborted = true
		return res
	}

	// Sacrifice ~half the sifted bits (deterministically: even indices) to
	// estimate QBER by public comparison.
	var mism, sampled int
	keepA := aSift[:0:0] // fresh slice, don't alias
	keepB := []int{}
	for i := range aSift {
		if i%2 == 0 {
			sampled++
			if aSift[i] != bSift[i] {
				mism++
			}
		} else {
			keepA = append(keepA, aSift[i])
			keepB = append(keepB, bSift[i])
		}
	}
	res.Sampled = sampled
	if sampled > 0 {
		res.QBER = float64(mism) / float64(sampled)
	}

	if res.QBER > QBERAbortThreshold {
		res.Aborted = true
		return res
	}

	// Surviving bits form the key; both sides must agree on every one.
	res.Agreement = equalInts(keepA, keepB)
	if !res.Agreement || len(keepA) == 0 {
		res.Aborted = true
		return res
	}
	res.KeyBits = keepA
	res.Key = DeriveKey(keepA)
	return res
}

// DeriveKey turns sifted key bits into a 32-byte key via SHA-256 (a stand-in for
// proper privacy amplification — it compresses the agreed bits into a uniform key).
func DeriveKey(bits []int) []byte {
	packed := make([]byte, (len(bits)+7)/8)
	for i, b := range bits {
		if b == 1 {
			packed[i/8] |= 1 << uint(i%8)
		}
	}
	sum := sha256.Sum256(packed)
	return sum[:]
}

func equalInts(a, b []int) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}
